### PR TITLE
pfconfig subcache amelioration

### PR DESCRIPTION
**Pull request by julsemaan**
# Description

Use the subcache in all relevant pfconfig::cached subclasses calls
Also do not constantly call pfconfig on undef key
Use Time::HiRes for the control file timestamp instead of POSIX
Added tests for subcache + subcache computing
# Impacts

Configuration in the processes (pfconfig clients), doesn't change the pfconfig server part
# Delete branch after merge

YES
# NEWS file entries
## Enhancements
- Added additional in-process caching for pfconfig proxied configuration

### CODE REVIEW

File: lib/pfconfig/cached.pm [hunk 1]
```diff
@@ -144,6 +145,35 @@ sub set_in_subcache {
...
+    elsif(defined($subcache_value)){
+        return $subcache_value;
+    }
+
+    my $result = $on_miss->();
+    if(defined($result)){
+        $self->set_in_subcache($key,$result);
+    }
+    else {
+        $self->set_in_subcache($key, pfconfig::undef_element->new);
```
Comment (jrouzierinverse) [inline lib/pfconfig/cached.pm]: This does not need to created every time.
This could be a singleton.

File: lib/pfconfig/cached_array.pm [hunk 2]
```diff
@@ -82,13 +82,11 @@ sub FETCH {
...
-
-    my $reply = $self->_get_from_socket("$self->{_namespace};$index");
-    my $result = defined($reply) ? $self->_get_from_socket("$self->{_namespace};$index")->{element} : undef;
-
-    $self->set_in_subcache( $index, $result );
+    my $result = $self->compute_from_subcache($index, sub {
+      my $reply = $self->_get_from_socket("$self->{_namespace};$index");
+      my $result = defined($reply) ? $reply->{element} : undef;
+      return $result;
+    });
```
Comment (jrouzierinverse) [inline lib/pfconfig/cached_array.pm:89]: I like the compute_from_subcache

File: lib/pfconfig/manager.pm [hunk 3]
```diff
@@ -189,9 +189,8 @@ sub touch_cache {
         close($fh);
     }
     if ( -e $filename ) {
-        sysopen( my $fh, $filename, O_RDWR | O_CREAT );
-        POSIX::2008::futimens( fileno $fh );
-        close($fh);
+        my $command = 'touch --date=@'.time.' '."'".$filename."'";
+        `$command`;
```
Comment (louismunro) [inline lib/pfconfig/manager.pm:193]: What was wrong with the old way?

File: t/unittest/pfconfig/cached.t [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+You should have received a copy of the GNU General Public License
+along with this program; if not, write to the Free Software
+Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301,
+USA.
+
+=cut
+
+1;
+
+
```
Comment (louismunro) [inline t/unittest/pfconfig/cached.t:138]: +1 for including unit tests.

**Comment by louismunro** [comment]
LGTM. Let's merge it.

File: lib/pfconfig/manager.pm (hunk 3 above)
Comment (extrafu) [inline lib/pfconfig/manager.pm:193]: Why not use Perl's utime? (http://perldoc.perl.org/functions/utime.html)

Comment (julsemaan) [inline lib/pfconfig/manager.pm:193]: The old way with POSIX doesn't seem to use the same source as Time::HiRes as it was nanoseconds late of the latest.

That meant the processes were being expired a few nanoseconds after which made unit tests fail.

I'll try utime and see if it works with subseconds

Comment (julsemaan) [inline lib/pfconfig/manager.pm:193]: Tested utime and it doesn't handle the microseconds.

**Pull request merged**